sbx/recurrent_ppo/policies.py

```diff
@@ -0,0 +1,415 @@
+import functools
+
+from dataclasses import field
+from typing import Any, Callable, Dict, List, Optional, Sequence, Union, Tuple
+
+import flax.linen as nn
+import gymnasium as gym
+import jax
+import jax.numpy as jnp
+import numpy as np
+import optax
+import tensorflow_probability.substrates.jax as tfp
+from flax.linen.initializers import constant
+from flax.training.train_state import TrainState
+from gymnasium import spaces
+from stable_baselines3.common.type_aliases import Schedule
+
+from sbx.common.policies import BaseJaxPolicy
+from sbx.common.recurrent import LSTMStates
+
+tfd = tfp.distributions
+
+
+# Added a ScanLSTM Module that automatically handles the reset of LSTM states
+# inspired from the ScanRNN in purejaxrl : https://github.com/luchris429/purejaxrl/blob/main/purejaxrl/ppo_rnn.py
+class ScanLSTM(nn.Module):
+    @functools.partial(
+        nn.scan,
+        variable_broadcast='params',
+        in_axes=0,
+        out_axes=0,
+        split_rngs={'params': False}
+    )
+    @nn.compact
+    def __call__(self, lstm_states, inputs_and_resets):
+        # pass the pi and vf lstm states, as well as the obs and the resets
+        input, resets = inputs_and_resets
+        hidden_state, cell_state = lstm_states
+
+        # create new lstm states to replace the old ones if reset is True
+        reset_lstm_states = self.initialize_carry(hidden_state.shape[0], hidden_state.shape[1])
+
+        # handle the reset of the hidden lstm states
+        hidden_state = jnp.where(
+            resets[:, np.newaxis],
+            reset_lstm_states[0],
+            hidden_state
+        )
+        # handle the reset of the cell lstm states
+        cell_state = jnp.where(
+            resets[:, np.newaxis],
+            reset_lstm_states[1],
+            cell_state
+        )
+
+        lstm_states = (hidden_state, cell_state)
+        hidden_size = lstm_states[0].shape[-1]
+        
+        new_lstm_states, output = nn.LSTMCell(features=hidden_size)(lstm_states, input)
+        return new_lstm_states, output
+    
+    @staticmethod
+    def initialize_carry(batch_size, hidden_size):
+        # Returns a tuple of lstm states (hidden and cell states)
+        return nn.LSTMCell(features=hidden_size).initialize_carry(
+            rng=jax.random.PRNGKey(0), input_shape=(batch_size, hidden_size)
```

Comment:
always the same rng, is that intented?

Reply:
if so, they can be precomputed, no?

Reply:
> always the same rng, is that intented?

I think it is, so the reset states are always the same (I borrowed this from [purejaxrl](https://github.com/luchris429/purejaxrl/blob/main/purejaxrl/ppo_rnn.py))

> if so, they can be precomputed, no?

Then yes it should be the case.

I'll ask a friend that knows well about lstm ppo in jax to be sure.